src/particles/pusher/PlasmaParticleAdvance.cpp

```diff
@@ -148,56 +116,190 @@ AdvancePlasmaParticles (PlasmaParticleContainer& plasma, const Fields & fields,
+
+                    doGatherShapeN<depos_order.value>(xp, yp, ExmByp, EypBxp, Ezp, Bxp, Byp,
                             Bzp, slice_arr, psi_comp, ez_comp, bx_comp, by_comp,
                             bz_comp, dx_inv, dy_inv, x_pos_offset, y_pos_offset);
 
-                        if (use_laser) {
-                            doLaserGatherShapeN<depos_order.value>(xp, yp, Aabssqp, AabssqDxp,
-                                AabssqDyp, a_arr, dx_inv, dy_inv, x_pos_offset, y_pos_offset);
-                        }
-                        // update force terms for a single particle
-                        const amrex::Real q = can_ionize ? ion_lev[ip] * charge : charge;
-                        UpdateForceTerms(uxp[ip], uyp[ip], psip[ip], ExmByp, EypBxp, Ezp, Bxp, Byp,
-                                         Bzp, Aabssqp, AabssqDxp, AabssqDyp, Fx1[ip], Fy1[ip],
-                                         Fux1[ip], Fuy1[ip], Fpsi1[ip], clightsq, phys_const, q,
-                                         mass);
+                    if (use_laser) {
+                        doLaserGatherShapeN<depos_order.value>(xp, yp, Aabssqp, AabssqDxp,
+                            AabssqDyp, a_arr, dx_inv, dy_inv, x_pos_offset, y_pos_offset);
                     }
 
-                    if (do_push)
-                    {
-                        // push a single particle
-                        PlasmaParticlePush(xp, yp, zp, uxp[ip], uyp[ip], psip[ip], x_prev[ip],
-                                           y_prev[ip], ux_temp[ip], uy_temp[ip], psi_temp[ip],
-                                           Fx1[ip], Fy1[ip], Fux1[ip], Fuy1[ip], Fpsi1[ip],
-                                           Fx2[ip], Fy2[ip], Fux2[ip], Fuy2[ip], Fpsi2[ip],
-                                           Fx3[ip], Fy3[ip], Fux3[ip], Fuy3[ip], Fpsi3[ip],
-                                           Fx4[ip], Fy4[ip], Fux4[ip], Fuy4[ip], Fpsi4[ip],
-                                           Fx5[ip], Fy5[ip], Fux5[ip], Fuy5[ip], Fpsi5[ip],
-                                           dz, temp_slice, ip, SetPosition, enforceBC );
+                    amrex::Real q_mass_clight_ratio = charge_mass_clight_ratio;
+                    if (ion_lev) {
+                        q_mass_clight_ratio *= ion_lev[ip];
                     }
-                    return;
+                    Bxp *= clight;
+                    Byp *= clight;
+                    Aabssqp *= 0.5_rt; // TODO: fix units of Aabssqp
+                    AabssqDxp *= 0.25_rt * me_clight_mass_ratio;
+                    AabssqDyp *= 0.25_rt * me_clight_mass_ratio;
+
+#ifndef HIPACE_USE_AB5_PUSH
+
+                    constexpr int nsub = 4;
+                    const amrex::Real sdz = dz/nsub;
+
+                    amrex::Real ux = ux_half_step[ip];
+                    amrex::Real uy = uy_half_step[ip];
+                    amrex::Real psi = psi_half_step[ip];
+
+                    for (int isub=0; isub<nsub; ++isub) {
```

Comment:
I added a few comments, could you check whether they make sense to you?
Also, it is not clear why the first momentum 1/2 push is over `nsub` while the second is over `nsub/2`. Did I miss something? Could you add comments to clarify?
```suggestion
                    // First 1/2 push in momentum
                    for (int isub=0; isub<nsub; ++isub) {
```

Reply:
The comments you suggested are wrong. The second momentum push is to convert from the leapfrog half step momentum to a full step one for current deposition, while the first momentum push does a full step.